lib/collection/array.js

```diff
@@ -158,33 +157,31 @@ export class ArrayColl extends KeyValueCollection {
   }
 
   removeKey(i) {
-    var item = this._array[i];
-    if (item == undefined) {
-      return;
+    if (i in this._array) {
+      this._notifyRemoved(this._array.splice(i, 1));
     }
-    delete this._array[i];
-    this._notifyRemoved([item]);
   }
 
   getKeyForValue(value) {
-    for (var i in this._array) {
-      if (this._array[i] == value) {
-        return i;
-      }
-    }
-    return undefined;
+    let key = this._array.indexOf(value);
+    return key < 0 ? undefined : key;
   }
 
   replaceAll(coll) {
-    // Call observers
-    super.replaceAll(coll);
+    let removedItems = this._array;
 
-    // Ensure same order and duplicity
-    this._array.length = 0;
     if (coll instanceof Collection) {
-      coll = coll.contents;
+      this._array = coll.contents;
+    } else {
+      this._array = coll.slice();
+    }
+    let addedItems = this._array.filter(item => !arrayRemove(removedItems, item));
```

Comment:
Smart, but hard to read code.
Side effect: changing removedItems while calculating addedItems.